### PR TITLE
[TG-8284] Expression simplification for indexOf()

### PR DESCRIPTION
**Pull request by danpoe**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/util/simplify_expr.cpp [hunk 1]
```diff
@@ -274,6 +274,98 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
...
+    }
+
+    const array_exprt &s1_data = s1_data_opt->get();
+
+    exprt::operandst ops;
+
+    exprt::operandst::const_iterator string_begin;
+    exprt::operandst::const_iterator string_end;
+
+    if(function_app.arguments().at(1).id() == ID_constant)
```
Comment (smowton) [inline src/util/simplify_expr.cpp]: Worth clarifying that this case is for index-of-char? Assert that the type of the constant is appropriate?

File: src/util/simplify_expr.cpp [hunk 2]
```diff
@@ -274,6 +274,98 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
...
+      string_end = ops.end();
+    }
+    else
+    {
+      const refined_string_exprt &s2 =
+        to_string_expr(function_app.arguments().at(1));
+
+      const auto s2_data_opt = try_get_string_data_array(s2, ns);
+
+      if(!s2_data_opt)
```
Comment (smowton) [inline src/util/simplify_expr.cpp]: Whereas this is index-of-string

**Review by smowton — APPROVED** [review]

File: src/util/simplify_expr.cpp [hunk 3]
```diff
@@ -274,6 +274,98 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
...
+      to_string_expr(function_app.arguments().at(0));
+
+    const auto s1_data_opt = try_get_string_data_array(s1, ns);
+
+    if(!s1_data_opt)
+    {
+      return true;
+    }
+
+    const array_exprt &s1_data = s1_data_opt->get();
```
Comment (romainbrenguier) [inline src/util/simplify_expr.cpp]: I'm not sure this is really useful

File: src/util/simplify_expr.cpp [hunk 4]
```diff
@@ -274,6 +274,98 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
...
+
+    exprt::operandst ops;
```
Comment (romainbrenguier) [inline src/util/simplify_expr.cpp]: Not clear here what this represents. It seems to be only used in the character case. I suggest separating the string and char cases, and maybe extract a function for the part that is common for the two.

File: src/util/simplify_expr.cpp [hunk 5]
```diff
@@ -274,6 +274,98 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
...
+
+    exprt::operandst::const_iterator string_begin;
+    exprt::operandst::const_iterator string_end;
```
Comment (romainbrenguier) [inline src/util/simplify_expr.cpp]: can maybe be replaced by `ranget<exprt::operandst::const_iterator>`

File: src/util/simplify_expr.cpp [hunk 6]
```diff
@@ -274,6 +274,98 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
...
+      string_end = s2_data.operands().end();
+    }
+
+    std::size_t starting_index = 0;
+
+    if(function_app.arguments().size() == 3)
+    {
+      auto &starting_index_expr = function_app.arguments().at(2);
+
+      if(starting_index_expr.id() != ID_constant)
```
Comment (romainbrenguier) [inline src/util/simplify_expr.cpp]: :pick: I'd prefer `if(const auto &c = try_dynamic_cast<constant_exprt>(function_app.arguments().at(2)))` and return true in the `else` case

File: src/util/simplify_expr.cpp [hunk 7]
```diff
@@ -274,6 +274,98 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
...
+
+      const array_exprt &s2_data = s2_data_opt->get();
+
+      string_begin = s2_data.operands().begin();
+      string_end = s2_data.operands().end();
+    }
+
+    std::size_t starting_index = 0;
+
+    if(function_app.arguments().size() == 3)
```
Comment (romainbrenguier) [inline src/util/simplify_expr.cpp]: :pick: I would wrap that code in an immediately interpreted lambda to make it clear that this is only used for computing the starting index

File: src/util/simplify_expr.cpp [hunk 8]
```diff
@@ -274,6 +274,98 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
...
+        string_begin,
+        string_end);
+
+      if(it == s1_data.operands().end())
+      {
+        expr = from_integer(-1, expr.type());
+      }
+      else
+      {
+        std::size_t idx = std::distance(s1_data.operands().begin(), it);
```
Comment (romainbrenguier) [inline src/util/simplify_expr.cpp]: const

File: src/util/simplify_expr.cpp [hunk 9]
```diff
@@ -274,6 +274,98 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
...
+    {
+      auto it = std::search(
+        std::next(s1_data.operands().begin(), starting_index),
+        s1_data.operands().end(),
+        string_begin,
+        string_end);
+
+      if(it == s1_data.operands().end())
+      {
+        expr = from_integer(-1, expr.type());
```
Comment (romainbrenguier) [inline src/util/simplify_expr.cpp]: :pick: I would make this `if` a ternary expression to assign expr

**Review by romainbrenguier — COMMENTED** [review]
I suggest defining a static function for the part that is common to char and string and clearly separate the two cases in simplify_function_application

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 053fc230553450341cce806769a3b13d21612f84).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116695086
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 089dd0cc751aa2fb46e6e74ad88b4f0d6f346f12).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116694980
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7b7d526bccefab324377e334f9bbd51cff560af7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118768374

**Pull request merged**